File: csharp/structural/rare_flyweight.cs
namespace Gof.ModernCSharp.Structural;

// Flyweight shares reusable intrinsic state across many small objects.
// Modern C# note: use only when profiling shows duplicate object memory pressure.
// Caches, string interning, records, and value types often cover the need.
public static class RareFlyweight
{
    sealed class Glyph(char value)
    {
        public string Draw((int X, int Y) position) => $"{value}@{position}";
    }

    sealed class GlyphFactory
    {
        readonly Dictionary<char, Glyph> _glyphs = [];
        public Glyph Get(char value) => _glyphs.TryGetValue(value, out var glyph)
            ? glyph
            : _glyphs[value] = new Glyph(value);
    }

    public static string Run()
    {
        var factory = new GlyphFactory();
        return ReferenceEquals(factory.Get('A'), factory.Get('A')).ToString();
    }
}

File: py/structural/rare_flyweight.py
# Flyweight shares reusable intrinsic state across many small objects.
# Modern Python note: use this only when profiling shows many duplicate objects or
# real memory pressure. Caches, interning, and value objects often cover the need.
class Glyph:
    def __init__(self, char):
        self.char = char

    def draw(self, position):
        return f"{self.char}@{position}"


class GlyphFactory:
    def __init__(self):
        self._glyphs = {}

    def get(self, char):
        if char not in self._glyphs:
            self._glyphs[char] = Glyph(char)
        return self._glyphs[char]


if __name__ == "__main__":
    factory = GlyphFactory()
    first = factory.get("A")
    second = factory.get("A")
    print(first is second)
    print(first.draw((10, 20)))
